File: awesome-calendar/src/main/java/com/archit/calendardaterangepicker/customviews/DateRangeMonthView.java
package com.archit.calendardaterangepicker.customviews;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.support.annotation.ColorInt;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import com.archit.calendardaterangepicker.R;
import com.archit.calendardaterangepicker.manager.DateRangeCalendarManager;
import com.archit.calendardaterangepicker.models.CalendarStyleAttr;
import com.archit.calendardaterangepicker.models.DayContainer;
import com.archit.calendardaterangepicker.timepicker.AwesomeTimePickerDialog;

import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;

/**
 * Created by archit.shah on 08/09/2017.
 */

public class DateRangeMonthView extends LinearLayout {

    private static final String LOG_TAG = DateRangeMonthView.class.getSimpleName();
    private Context mContext;
    private LinearLayout llDaysContainer;
    private LinearLayout llTitleWeekContainer;

    private Calendar currentCalendarMonth;

    private CalendarStyleAttr calendarStyleAttr;

    private static final PorterDuff.Mode drawableMode = PorterDuff.Mode.SRC_OVER;


    private DateRangeCalendarView.CalendarListener calendarListener;

    private DateRangeCalendarManager dateRangeCalendarManager;

    public DateRangeCalendarView.CalendarListener getCalendarListener() {
        return calendarListener;
    }

    public void setCalendarListener(DateRangeCalendarView.CalendarListener calendarListener) {
        this.calendarListener = calendarListener;
    }

    public DateRangeMonthView(Context context) {
        super(context);
        initView(context, null);
    }

    public DateRangeMonthView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context, attrs);
    }

    public DateRangeMonthView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context, attrs);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public DateRangeMonthView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initView(context, attrs);
    }

    /**
     * To initialize child views
     *
     * @param context      - App context
     * @param attributeSet - Attr set
     */
    private void initView(Context context, AttributeSet attributeSet) {
        mContext = context;

        LayoutInflater layoutInflater = LayoutInflater.from(mContext);
        LinearLayout mainView = (LinearLayout) layoutInflater.inflate(R.layout.layout_calendar_month, this, true);
        llDaysContainer = (LinearLayout) mainView.findViewById(R.id.llDaysContainer);
        llTitleWeekContainer = (LinearLayout) mainView.findViewById(R.id.llTitleWeekContainer);

        setListeners();

        if (isInEditMode()) {
            return;
        }

    }



    /**
     * To set listeners.
     */
    private void setListeners() {


    }


    private OnClickListener dayClickListener = new OnClickListener() {
        @Override
        public void onClick(View view) {

            int key = (int) view.getTag();
            final Calendar selectedCal = Calendar.getInstance();
            Date date = new Date();
            try {
                date = DateRangeCalendarManager.SIMPLE_DATE_FORMAT.parse(String.valueOf(key));
            } catch (ParseException e) {
                e.printStackTrace();
            }
            selectedCal.setTime(date);

            Calendar minSelectedDate = dateRangeCalendarManager.getMinSelectedDate();
            Calendar maxSelectedDate = dateRangeCalendarManager.getMaxSelectedDate();

            if (minSelectedDate != null && maxSelectedDate == null) {
                maxSelectedDate = selectedCal;

                int startDateKey = DayContainer.GetContainerKey(minSelectedDate);
                int lastDateKey = DayContainer.GetContainerKey(maxSelectedDate);

                if (startDateKey == lastDateKey) {
                    minSelectedDate = maxSelectedDate;
                } else if (startDateKey > lastDateKey) {
                    Calendar temp = (Calendar) minSelectedDate.clone();
                    minSelectedDate = maxSelectedDate;
                    maxSelectedDate = temp;
                }
            } else if (maxSelectedDate == null) {
                //This will call one time only
                minSelectedDate = selectedCal;
            } else {
                minSelectedDate = selectedCal;
                maxSelectedDate = null;
            }

            dateRangeCalendarManager.setMinSelectedDate(minSelectedDate);
            dateRangeCalendarManager.setMaxSelectedDate(maxSelectedDate);
            drawCalendarForMonth(currentCalendarMonth);

            if (calendarStyleAttr.isShouldEnabledTime()) {
                final Calendar finalMinSelectedDate = minSelectedDate;
                final Calendar finalMaxSelectedDate = maxSelectedDate;
                AwesomeTimePickerDialog awesomeTimePickerDialog = new AwesomeTimePickerDialog(mContext, mContext.getString(R.string.select_time), new AwesomeTimePickerDialog.TimePickerCallback() {
                    @Override
                    public void onTimeSelected(int hours, int mins) {
                        selectedCal.set(Calendar.HOUR, hours);
                        selectedCal.set(Calendar.MINUTE, mins);

                        Log.i(LOG_TAG, "Time: " + selectedCal.getTime().toString());
                        if (calendarListener != null) {

                            if (finalMaxSelectedDate != null) {
                                calendarListener.onDateRangeSelected(finalMinSelectedDate, finalMaxSelectedDate);
                            } else {
                                calendarListener.onFirstDateSelected(finalMinSelectedDate);
                            }
                        }
                    }

                    @Override
                    public void onCancel() {
                        DateRangeMonthView.this.resetAllSelectedViews();
                    }
                });
                awesomeTimePickerDialog.showDialog();
            } else {
                Log.i(LOG_TAG, "Time: " + selectedCal.getTime().toString());
                if (maxSelectedDate != null) {
                    calendarListener.onDateRangeSelected(minSelectedDate, maxSelectedDate);
                } else {
                    calendarListener.onFirstDateSelected(minSelectedDate);
                }
            }
        }
    };

    /**
     * To draw calendar for the given month. Here calendar object should start from date of 1st.
     *
     * @param month
     */
    public void drawCalendarForMonth(CalendarStyleAttr calendarStyleAttr, Calendar month, DateRangeCalendarManager dateRangeCalendarManager) {
        this.calendarStyleAttr = calendarStyleAttr;
        this.currentCalendarMonth = (Calendar) month.clone();
        this.dateRangeCalendarManager = dateRangeCalendarManager;
        setFonts();
        setWeekTitleColor(calendarStyleAttr.getWeekColor());
        drawCalendarForMonth(currentCalendarMonth);
    }

    /**
     * To draw calendar for the given month. Here calendar object should start from date of 1st.
     *
     * @param month
     */
    private void drawCalendarForMonth(Calendar month) {

        Log.v(LOG_TAG, "Current cal: " + month.getTime().toString());
        currentCalendarMonth = (Calendar) month.clone();
        int startDay = month.get(Calendar.DAY_OF_WEEK);


        month.add(Calendar.DATE, -startDay + 1);

        for (int i = 0; i < llDaysContainer.getChildCount(); i++) {
            LinearLayout weekRow = (LinearLayout) llDaysContainer.getChildAt(i);

            for (int j = 0; j < 7; j++) {
                RelativeLayout rlDayContainer = (RelativeLayout) weekRow.getChildAt(j);

                DayContainer container = new DayContainer(rlDayContainer);

                container.tvDate.setText(String.valueOf(month.get(Calendar.DATE)));
                if (calendarStyleAttr.getFonts() != null) {
                    container.tvDate.setTypeface(calendarStyleAttr.getFonts());
                }
                Log.v(LOG_TAG, "Date: " + month.getTime().toString());
                drawDayContainer(container, month);
                month.add(Calendar.DATE, 1);
            }
        }
    }

    /**
     * To draw specific date container according to past date, today, selected or from range.
     *
     * @param container - Date container
     * @param calendar  - Calendar obj of specific date of the month.
     */
    private void drawDayContainer(DayContainer container, Calendar calendar) {

        Calendar today = Calendar.getInstance();

        int date = calendar.get(Calendar.DATE);

        if (currentCalendarMonth.get(Calendar.MONTH) != calendar.get(Calendar.MONTH)) {
            hideDayContainer(container);
        } else if (today.after(calendar) && (today.get(Calendar.DAY_OF_YEAR) != calendar.get(Calendar.DAY_OF_YEAR))) {
            disableDayContainer(container);
            container.tvDate.setText(String.valueOf(date));
        } else {

            @DateRangeCalendarManager.RANGE_TYPE
            int type = dateRangeCalendarManager.checkDateRange(calendar);
            if (type == DateRangeCalendarManager.RANGE_TYPE.START_DATE || type == DateRangeCalendarManager.RANGE_TYPE.LAST_DATE) {
                makeAsSelectedDate(container, type);
            } else if (type == DateRangeCalendarManager.RANGE_TYPE.MIDDLE_DATE) {
                makeAsRangeDate(container);
            } else {
                enabledDayContainer(container);
            }

            container.tvDate.setText(String.valueOf(date));
        }

        container.rootView.setTag(DayContainer.GetContainerKey(calendar));
    }

    /**
     * To hide date if date is from previous month.
     *
     * @param container - Container
     */
    private void hideDayContainer(DayContainer container) {
        container.tvDate.setText("");
        container.tvDate.setBackgroundColor(Color.TRANSPARENT);
        container.strip.setBackgroundColor(Color.TRANSPARENT);
        container.rootView.setBackgroundColor(Color.TRANSPARENT);
        container.rootView.setVisibility(INVISIBLE);
        container.rootView.setOnClickListener(null);
    }

    /**
     * To disable past date. Click listener will be removed.
     *
     * @param container - Container
     */
    private void disableDayContainer(DayContainer container) {
        container.tvDate.setBackgroundColor(Color.TRANSPARENT);
        container.strip.setBackgroundColor(Color.TRANSPARENT);
        container.rootView.setBackgroundColor(Color.TRANSPARENT);
        container.tvDate.setTextColor(calendarStyleAttr.getDisableDateColor());
        container.rootView.setVisibility(VISIBLE);
        container.rootView.setOnClickListener(null);
    }

    /**
     * To enable date by enabling click listeners.
     *
     * @param container - Container
     */
    private void enabledDayContainer(DayContainer container) {
        container.tvDate.setBackgroundColor(Color.TRANSPARENT);
        container.strip.setBackgroundColor(Color.TRANSPARENT);
        container.rootView.setBackgroundColor(Color.TRANSPARENT);
        container.tvDate.setTextColor(calendarStyleAttr.getDefaultDateColor());
        container.rootView.setVisibility(VISIBLE);
        container.rootView.setOnClickListener(dayClickListener);
    }

    /**
     * To draw date container as selected as end selection or middle selection.
     *
     * @param container - Container
     * @param stripType - Right end date, Left end date or middle
     */
    private void makeAsSelectedDate(DayContainer container,@DateRangeCalendarManager.RANGE_TYPE int stripType) {
        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) container.strip.getLayoutParams();

        Calendar maxDate = dateRangeCalendarManager.getMaxSelectedDate();

        if (stripType == DateRangeCalendarManager.RANGE_TYPE.START_DATE && maxDate != null) {
            GradientDrawable mDrawable = (GradientDrawable) ContextCompat.getDrawable(mContext, R.drawable.range_bg_left);
            mDrawable.setColor(calendarStyleAttr.getRangeStripColor());
            container.strip.setBackground(mDrawable);
            layoutParams.setMargins(20, 0, 0, 0);
        } else if (stripType == DateRangeCalendarManager.RANGE_TYPE.LAST_DATE) {
            GradientDrawable mDrawable = (GradientDrawable) ContextCompat.getDrawable(mContext, R.drawable.range_bg_right);
            mDrawable.setColor(calendarStyleAttr.getRangeStripColor());
            container.strip.setBackground(mDrawable);
            layoutParams.setMargins(0, 0, 20, 0);
        } else {
            container.strip.setBackgroundColor(Color.TRANSPARENT);
            layoutParams.setMargins(0, 0, 0, 0);
        }
        container.strip.setLayoutParams(layoutParams);
        GradientDrawable mDrawable = (GradientDrawable) ContextCompat.getDrawable(mContext, R.drawable.green_circle);
        mDrawable.setColor(calendarStyleAttr.getSelectedDateCircleColor());
        container.tvDate.setBackground(mDrawable);
        container.rootView.setBackgroundColor(Color.TRANSPARENT);
        container.tvDate.setTextColor(calendarStyleAttr.getSelectedDateColor());
        container.rootView.setVisibility(VISIBLE);
        container.rootView.setOnClickListener(dayClickListener);
    }

    /**
     * To draw date as middle date
     *
     * @param container - Container
     */
    private void makeAsRangeDate(DayContainer container) {
        container.tvDate.setBackgroundColor(Color.TRANSPARENT);
        GradientDrawable mDrawable = (GradientDrawable) ContextCompat.getDrawable(mContext, R.drawable.range_bg);
        mDrawable.setColor(calendarStyleAttr.getRangeStripColor());
        container.strip.setBackground(mDrawable);
        container.rootView.setBackgroundColor(Color.TRANSPARENT);
        container.tvDate.setTextColor(calendarStyleAttr.getRangeDateColor());
        container.rootView.setVisibility(VISIBLE);
        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) container.strip.getLayoutParams();
        layoutParams.setMargins(0, 0, 0, 0);
        container.strip.setLayoutParams(layoutParams);
        container.rootView.setOnClickListener(dayClickListener);
    }


    /**
     * To remove all selection and redraw current calendar
     */
    public void resetAllSelectedViews() {

        dateRangeCalendarManager.setMinSelectedDate(null);
        dateRangeCalendarManager.setMaxSelectedDate(null);

        drawCalendarForMonth(currentCalendarMonth);

    }


    /**
     * To set week title color
     *
     * @param color - resource color value
     */
    public void setWeekTitleColor(@ColorInt int color) {
        for (int i = 0; i < llTitleWeekContainer.getChildCount(); i++) {
            CustomTextView textView = (CustomTextView) llTitleWeekContainer.getChildAt(i);
            textView.setTextColor(color);
        }
    }

    /**
     * To apply custom fonts to all the text views
     */
    private void setFonts() {

        drawCalendarForMonth(currentCalendarMonth);

        for (int i = 0; i < llTitleWeekContainer.getChildCount(); i++) {

            CustomTextView textView = (CustomTextView) llTitleWeekContainer.getChildAt(i);
            textView.setTypeface(calendarStyleAttr.getFonts());

        }
    }
}

File: awesome-calendar/src/main/java/com/archit/calendardaterangepicker/customviews/DateRangeMonthView.kt
package com.archit.calendardaterangepicker.customviews

import android.annotation.TargetApi
import android.content.Context
import android.os.Build.VERSION_CODES
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import com.archit.calendardaterangepicker.R
import com.archit.calendardaterangepicker.R.array
import com.archit.calendardaterangepicker.R.layout
import com.archit.calendardaterangepicker.R.string
import com.archit.calendardaterangepicker.customviews.CalendarDateRangeManager.DateSelectionState.IN_SELECTED_RANGE
import com.archit.calendardaterangepicker.customviews.CalendarDateRangeManager.DateSelectionState.LAST_DATE
import com.archit.calendardaterangepicker.customviews.CalendarDateRangeManager.DateSelectionState.START_DATE
import com.archit.calendardaterangepicker.customviews.CalendarDateRangeManager.DateSelectionState.START_END_SAME
import com.archit.calendardaterangepicker.customviews.DateTiming.NONE
import com.archit.calendardaterangepicker.customviews.DateView.Companion.getContainerKey
import com.archit.calendardaterangepicker.customviews.DateView.DateState
import com.archit.calendardaterangepicker.customviews.DateView.DateState.DISABLE
import com.archit.calendardaterangepicker.customviews.DateView.DateState.END
import com.archit.calendardaterangepicker.customviews.DateView.DateState.HIDDEN
import com.archit.calendardaterangepicker.customviews.DateView.DateState.MIDDLE
import com.archit.calendardaterangepicker.customviews.DateView.DateState.SELECTABLE
import com.archit.calendardaterangepicker.customviews.DateView.DateState.START
import com.archit.calendardaterangepicker.customviews.DateView.OnDateClickListener
import com.archit.calendardaterangepicker.models.CalendarStyleAttributes
import com.archit.calendardaterangepicker.models.CalendarStyleAttributes.DateSelectionMode.FIXED_RANGE
import com.archit.calendardaterangepicker.models.CalendarStyleAttributes.DateSelectionMode.FREE_RANGE
import com.archit.calendardaterangepicker.models.CalendarStyleAttributes.DateSelectionMode.SINGLE
import com.archit.calendardaterangepicker.timepicker.AwesomeTimePickerDialog
import com.archit.calendardaterangepicker.timepicker.AwesomeTimePickerDialog.TimePickerCallback
import java.util.Calendar

/**
 * Created by archit.shah on 08/09/2017.
 */
internal class DateRangeMonthView : LinearLayout {
    private lateinit var llDaysContainer: LinearLayout
    private lateinit var llTitleWeekContainer: LinearLayout
    private lateinit var currentCalendarMonth: Calendar
    private lateinit var calendarStyleAttr: CalendarStyleAttributes
    private var calendarListener: CalendarListener? = null
    private lateinit var dateRangeCalendarManager: CalendarDateRangeManager
    fun setCalendarListener(calendarListener: CalendarListener?) {
        this.calendarListener = calendarListener
    }

    constructor(context: Context) : super(context) {
        initView(context, null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initView(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initView(context, attrs)
    }

    @TargetApi(VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int,
                defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        initView(context, attrs)
    }

    /**
     * To initialize child views
     *
     * @param context      - App context
     * @param attributeSet - Attr set
     */
    private fun initView(context: Context, attributeSet: AttributeSet?) {
        val layoutInflater = LayoutInflater.from(context)
        val mainView = layoutInflater.inflate(layout.layout_calendar_month, this, true) as LinearLayout
        llDaysContainer = mainView.findViewById(R.id.llDaysContainer)
        llTitleWeekContainer = mainView.findViewById(R.id.llTitleWeekContainer)
        setListeners()
    }

    /**
     * To set listeners.
     */
    private fun setListeners() {}
    private val mOnDateClickListener: OnDateClickListener = object : OnDateClickListener {
        override fun onDateClicked(view: View, selectedDate: Calendar) {
            if (calendarStyleAttr.isEditable) {
                if (calendarStyleAttr.isShouldEnabledTime) {
                    val awesomeTimePickerDialog = AwesomeTimePickerDialog(context,
                            context.getString(string.select_time), object : TimePickerCallback {
                        override fun onTimeSelected(hours: Int, mins: Int) {
                            selectedDate[Calendar.HOUR] = hours
                            selectedDate[Calendar.MINUTE] = mins
                            setSelectedDate(selectedDate)
                        }

                        override fun onCancel() {
                            resetAllSelectedViews()
                        }
                    })
                    awesomeTimePickerDialog.showDialog()
                } else {
                    setSelectedDate(selectedDate)
                }
            }
        }
    }

    private fun setSelectedDate(selectedDate: Calendar) {

        val selectionMode = calendarStyleAttr.dateSelectionMode
        var minSelectedDate = dateRangeCalendarManager.getMinSelectedDate()
        var maxSelectedDate = dateRangeCalendarManager.getMaxSelectedDate()

        when (selectionMode) {
            FREE_RANGE -> {
                if (minSelectedDate != null && maxSelectedDate == null) {
                    maxSelectedDate = selectedDate
                    val startDateKey = getContainerKey(minSelectedDate)
                    val lastDateKey = getContainerKey(maxSelectedDate)
                    if (startDateKey == lastDateKey) {
                        minSelectedDate = maxSelectedDate
                    } else if (startDateKey > lastDateKey) {
                        val temp = minSelectedDate.clone() as Calendar
                        minSelectedDate = maxSelectedDate
                        maxSelectedDate = temp
                    }
                } else if (maxSelectedDate == null) {
                    //This will call one time only
                    minSelectedDate = selectedDate
                } else {
                    minSelectedDate = selectedDate
                    maxSelectedDate = null
                }
            }
            SINGLE -> {
                minSelectedDate = selectedDate
                maxSelectedDate = selectedDate
            }
            FIXED_RANGE -> {
                minSelectedDate = selectedDate
                maxSelectedDate = selectedDate.clone() as Calendar
                maxSelectedDate.add(Calendar.DATE, calendarStyleAttr.fixedDaysSelectionNumber)
            }
        }

        dateRangeCalendarManager.setSelectedDateRange(minSelectedDate, maxSelectedDate)
        drawCalendarForMonth(currentCalendarMonth)
        Log.i(LOG_TAG, "Time: " + selectedDate.time.toString())
        if (maxSelectedDate != null) {
            calendarListener!!.onDateRangeSelected(minSelectedDate, maxSelectedDate)
        } else {
            calendarListener!!.onFirstDateSelected(minSelectedDate)
        }
    }

    /**
     * To draw calendar for the given month. Here calendar object should start from date of 1st.
     *
     * @param calendarStyleAttr        Calendar style attributes
     * @param month                    Month to be drawn
     * @param dateRangeCalendarManager Calendar data manager
     */
    fun drawCalendarForMonth(calendarStyleAttr: CalendarStyleAttributes, month: Calendar, dateRangeCalendarManager: CalendarDateRangeManager) {
        this.calendarStyleAttr = calendarStyleAttr
        currentCalendarMonth = month.clone() as Calendar
        this.dateRangeCalendarManager = dateRangeCalendarManager
        drawCalendarForMonth(currentCalendarMonth)
    }

    /**
     * To draw calendar for the given month. Here calendar object should start from date of 1st.
     *
     * @param month Calendar month
     */
    private fun drawCalendarForMonth(month: Calendar) {
        setWeekTextAttributes()
        currentCalendarMonth = month.clone() as Calendar
        currentCalendarMonth[Calendar.DATE] = 1
        resetTime(currentCalendarMonth, NONE)
        val weekTitle = context.resources.getStringArray(array.week_sun_sat)

        //To set week day title as per offset
        for (i in 0..6) {
            val textView = llTitleWeekContainer.getChildAt(i) as CustomTextView
            val weekStr = weekTitle[(i + calendarStyleAttr.weekOffset) % 7]
            textView.text = weekStr
        }
        var startDay = month[Calendar.DAY_OF_WEEK] - calendarStyleAttr.weekOffset

        //To rotate week day according to offset
        if (startDay < 1) {
            startDay = startDay + 7
        }
        month.add(Calendar.DATE, -startDay + 1)
        for (i in 0 until llDaysContainer.childCount) {
            val weekRow = llDaysContainer.getChildAt(i) as LinearLayout
            for (j in 0..6) {
                val customDateView = weekRow.getChildAt(j) as CustomDateView
                drawDayContainer(customDateView, month)
                month.add(Calendar.DATE, 1)
            }
        }
    }

    /**
     * To draw specific date container according to past date, today, selected or from range.
     *
     * @param customDateView - Date container
     * @param date       - Calendar obj of specific date of the month.
     */
    private fun drawDayContainer(customDateView: CustomDateView, date: Calendar) {
        customDateView.setDateText(date[Calendar.DATE].toString())
        customDateView.setDateStyleAttributes(calendarStyleAttr)
        customDateView.setDateClickListener(mOnDateClickListener)
        calendarStyleAttr.fonts?.let { customDateView.setTypeface(it) }
        val dateState: DateState
        dateState = if (currentCalendarMonth[Calendar.MONTH] != date[Calendar.MONTH]) {
            HIDDEN
        } else {
            val type = dateRangeCalendarManager.checkDateRange(date)
            if (type === START_DATE) {
                START
            } else if (type === LAST_DATE) {
                END
            } else if (type === START_END_SAME) {
                DateState.START_END_SAME
            } else if (type === IN_SELECTED_RANGE) {
                MIDDLE
            } else {
                if (dateRangeCalendarManager.isSelectableDate(date)) {
                    SELECTABLE
                } else {
                    DISABLE
                }
            }
        }
        customDateView.updateDateBackground(dateState)
        customDateView.tag = getContainerKey(date)
    }

    /**
     * To remove all selection and redraw current calendar
     */
    fun resetAllSelectedViews() {
        dateRangeCalendarManager.resetSelectedDateRange()
        drawCalendarForMonth(currentCalendarMonth)
    }

    /**
     * To apply configs to all the text views
     */
    private fun setWeekTextAttributes() {
        for (i in 0 until llTitleWeekContainer.childCount) {
            val textView = llTitleWeekContainer.getChildAt(i) as CustomTextView
            textView.typeface = calendarStyleAttr.fonts
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, calendarStyleAttr.textSizeWeek)
            textView.setTextColor(calendarStyleAttr.weekColor)
        }
    }

    companion object {
        private val LOG_TAG = DateRangeMonthView::class.java.simpleName
    }
}